### PR TITLE
fix greedy_modularity when multiple components exist.

### LINKED ISSUE #5530: greedy_modularity_communities raises StopIteration for unconnected (?) graph
After an update of `networkx` from 2.6.3 to 2.8, one of my tests suddenly failed with
```
Traceback (most recent call last):
  File "/usr/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/home/gereon/Develop/lexedata/src/lexedata/report/nonconcatenative_morphemes.py", line 159, in <module>
    networkx.algorithms.community.greedy_modularity_communities(graph),
  File "/home/gereon/.local/etc/lexedata/lib/python3.10/site-packages/networkx/algorithms/community/modularity_max.py", line 354, in greedy_modularity_communities
    dq = next(community_gen)
StopIteration

```
It turns out that the internals of `greedy_modularity_communities` have changed, so the function doesn't work any more with unconnected graphs.

### Current Behavior

`networkx.algorithms.community.greedy_modularity_communities` fails with undocumented error message for unconnected graphs, or graphs without edges

### Expected Behavior

 - Either document the requirement to have a connected graph with at least one edge and throw `ValueError` instead of `StopIteration` when it is not fulfilled, or run the graph through `nx.connected_components()` first.
 - If the graph has only one node and no edges, return that node.

### Steps to Reproduce

```
>>> G = nx.Graph()
>>> G.add_edges_from([(0,1),(2,3)])
>>> greedy_modularity_communities(G)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/gereon/.local/etc/lexedata/lib/python3.10/site-packages/networkx/algorithms/community/modularity_max.py", line 354, in greedy_modularity_communities
    dq = next(community_gen)
StopIteration
```

```
>>> G = nx.Graph()
>>> G.add_nodes_from([1])
>>> greedy_modularity_communities(G)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/gereon/.local/etc/lexedata/lib/python3.10/site-packages/networkx/algorithms/community/modularity_max.py", line 350, in greedy_modularity_communities
    communities = next(community_gen)
  File "/home/gereon/.local/etc/lexedata/lib/python3.10/site-packages/networkx/algorithms/community/modularity_max.py", line 79, in _greedy_modularity_communities_generator
    q0 = 1 / m
ZeroDivisionError: division by zero

```
### Environment

Python version: 3.10.2
NetworkX version: 2.8

### PR DESCRIPTION
**Pull request by dschult**
Fixes #5530 

After the community merging iteration stops, we have to check whether the number of communities is still too many for the `best_n` criteria.  If so, merge the biggest 2 components repeatedly to keep modularity as big as we can.

### CODE REVIEW

File: networkx/algorithms/community/tests/test_modularity_max.py
Comment (rossbar) [inline networkx/algorithms/community/tests/test_modularity_max.py:52]:
```suggestion
    # best_n can make the algorithm continue even when modularity goes down
```

File: networkx/algorithms/community/modularity_max.py
Comment (rossbar) [inline networkx/algorithms/community/modularity_max.py:361]: I have no idea what makes the most sense in this case, so I'd defer to the graph/network experts!

If no one's entirely sure what should happen here, one option could be to simply return the `communities` in the `best_n=None` (default) case and punt when `best_n!=None`, e.g. something like:

```python
# StopIteration occurs when communities are the connected components
except StopIteration:
    if best_n was None:  # Would need a variable to keep track of the original value of `best_n`
        return communities
    else:
        raise <exception-with-better-message-than-stop-iteration>
```
At least that would fix the returning-the-components regression part of the issue without having to necessarily make a decision on what to do with `best_n`.

**Review by rossbar — APPROVED** [review]
LGTM - I don't have any valuable input about what makes sense in the case when there's a disconnected graph and `best_n` is less than the number of components, so I leave that discussion to others.

Everything else LGTM though, so I approve pending the `best_n` decision.

File: networkx/algorithms/community/modularity_max.py
Comment (dschult) [reply networkx/algorithms/community/modularity_max.py:361]: Luckily, the only value of `best_n` that would represent an `original_best_n is None` is when `best_n == len(G)`.
And in that case, the following while loop is passed over and we return `communities` just as in your suggestion.

I think the question mark in the comment on this line (361) should be removed -- and made more clear.
Once we include a parameter `best_n` we have to expect that sometimes we will merge communities that decrease the modularity. The parameter indicates to keep merging until we get `best_n` communities.
How about a change for that comment to:
```#  if best_n requires more merging, merge the biggest to keep modularity high```
and then can remove comment 3 lines down.

File: networkx/algorithms/community/modularity_max.py
Comment (MridulS) [inline networkx/algorithms/community/modularity_max.py:330]: This is maybe not the right place but while reviewing this I realized that currently `greedy_modularity_communities` fails with a cryptic `ValueError: cutoff must be between 1 and 0. Got {something between 1 and 0}.` when a graph with no nodes is passed and it pops up with a `ZeroDivisionError` if there are no edges. We should probably catch them and make it more graceful.

I can create a new issue for this :)

Comment (dschult) [reply networkx/algorithms/community/modularity_max.py:330]: Yes!  Those corner cases certainly are not covered here.  
And there many be others...  3-nodes and 1 edge?

**Review by MridulS — APPROVED** [review]
I'll go ahead and create a new issue for the edge cases. We can merge this.

I'll wait if you would like to add anything else here, if not please merge this :)

**Pull request merged**
